sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,845 @@
+
+    See also
+    --------
+    sklearn.preprocessing.OrdinalEncoder : performs an ordinal (integer)
+      encoding of the categorical features.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    sklearn.preprocessing.LabelBinarizer : binarizes labels in a one-vs-all
+      fashion.
+    sklearn.preprocessing.MultiLabelBinarizer : transforms between iterable of
+      iterables and a multilabel format, e.g. a (samples x classes) binary
+      matrix indicating the presence of a class label.
+    """
+
+    def __init__(self, n_values=None, categorical_features=None,
+                 categories=None, sparse=True, dtype=np.float64,
+                 handle_unknown='error'):
+        self._categories = self.categories = categories
+        self.sparse = sparse
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+        self._n_values = self.n_values = n_values
+        self._categorical_features = categorical_features
+        self.categorical_features = categorical_features
+
+    # Deprecated attributes
+
+    @property
+    @deprecated("The 'active_features_' attribute was deprecated in version "
+                "0.20 and will be removed 0.22.")
+    def active_features_(self):
+        check_is_fitted(self, 'categories_')
+        return self._active_features_
+
+    @property
+    @deprecated("The 'feature_indices_' attribute was deprecated in version "
+                "0.20 and will be removed 0.22.")
+    def feature_indices_(self):
+        check_is_fitted(self, 'categories_')
+        return self._feature_indices_
+
+    @property
+    @deprecated("The 'n_values_' attribute was deprecated in version "
+                "0.20 and will be removed 0.22.")
+    def n_values_(self):
+        check_is_fitted(self, 'categories_')
+        return self._n_values_
+
+    def _handle_deprecations(self, X):
+
+        # user manually set the categories -> never legacy mode
+        if self._categories is not None:
+            self._legacy_mode = False
+
+        # categories not set -> infer if we need legacy mode or not
+        elif self._n_values is not None and self._n_values != 'auto':
+            msg = (
+                "Passing 'n_values' is deprecated in version 0.20 and will be "
+                "removed in 0.22. You can use the 'categories' keyword "
+                "instead. 'n_values=n' corresponds to 'categories=[range(n)]'."
+            )
+            warnings.warn(msg, DeprecationWarning)
+
+            # we internally translate this to the correct categories
+            # and don't use legacy mode
+            X = check_array(X, dtype=np.int)
+
+            if isinstance(self.n_values, numbers.Integral):
+                n_features = X.shape[1]
+                self.categories = [
```

Comment:
We conventionally avoid modifying constructor attributes in fit. Use private attributes if you need to...

Reply:
changed to use private attributes